zebrad/src/components/miner.rs

```diff
@@ -0,0 +1,527 @@
+            Response = zebra_state::ReadResponse,
+            Error = zebra_state::BoxError,
+        > + Clone
+        + Send
+        + Sync
+        + 'static,
+    <State as Service<zebra_state::ReadRequest>>::Future: Send,
+    Tip: ChainTip + Clone + Send + Sync + 'static,
+    BlockVerifierRouter: Service<zebra_consensus::Request, Response = block::Hash, Error = zebra_consensus::BoxError>
+        + Clone
+        + Send
+        + Sync
+        + 'static,
+    <BlockVerifierRouter as Service<zebra_consensus::Request>>::Future: Send,
+    SyncStatus: ChainSyncStatus + Clone + Send + Sync + 'static,
+    AddressBook: AddressBookPeers + Clone + Send + Sync + 'static,
+{
+    // Pass the correct arguments, even if Zebra currently ignores them.
+    let mut parameters = get_block_template::JsonParameters {
+        mode: Template,
+        data: None,
+        capabilities: vec![LongPoll, CoinbaseTxn],
+        long_poll_id: None,
+        _work_id: None,
+    };
+
+    // Shut down the task when all the template receivers are dropped, or Zebra shuts down.
+    while !template_sender.is_closed() && !is_shutting_down() {
+        let template = rpc.get_block_template(Some(parameters.clone())).await;
+
+        // Wait for the chain to sync so we get a valid template.
+        let Ok(template) = template else {
+            debug!(
+                ?BLOCK_TEMPLATE_WAIT_TIME,
+                "waiting for a valid block template",
+            );
+
+            // Skip the wait if we got an error because we are shutting down.
+            if !is_shutting_down() {
+                sleep(BLOCK_TEMPLATE_WAIT_TIME).await;
+            }
+
+            continue;
+        };
+
+        // Convert from RPC GetBlockTemplate to Block
+        let template = template
+            .try_into_template()
+            .expect("invalid RPC response: proposal in response to a template request");
+
+        info!(
+            height = ?template.height,
+            transactions = ?template.transactions.len(),
+            "mining with an updated block template",
+        );
+
+        // Tell the next get_block_template() call to wait until the template has changed.
+        parameters.long_poll_id = Some(template.long_poll_id);
+
+        let block = proposal_block_from_template(&template, TimeSource::CurTime)
+            .expect("unexpected invalid block template");
+
+        // If the template has actually changed, send an updated template.
+        template_sender.send_if_modified(|old_block| {
+            if old_block.as_ref().map(|b| *b.header) == Some(*block.header) {
+                return false;
+            }
+            *old_block = Some(Arc::new(block));
+            true
+        });
+
+        // If the blockchain is changing rapidly, limit how often we'll update the template.
+        // But if we're shutting down, do that immediately.
+        if !template_sender.is_closed() && !is_shutting_down() {
+            sleep(Duration::from_secs(
```

Comment:
Optional: Put this in a join so it's awaited in parallel to the long poll result.

Reply:
Can you explain why this change would be helpful? Is the difference important enough to complicate the code?

It is usually clearer to split code into another function, rather than assigning an async block to a named variable.

Reply:
> Can you explain why this change would be helpful?

If a long polling request returns right before a new block is added to the chain, it could delay mining by a few seconds.

I think we want to check if that the new template's previous block hash is the same as before, or that it's been longer than `GET_BLOCK_TEMPLATE_MEMPOOL_LONG_POLL_INTERVAL` since a new template was sent to the watch channel?

> Is the difference important enough to complicate the code?

I'm hoping it'll be relatively simple, but otherwise no, it's a marginal efficiency gain, and this intended to be used for testing.

> It is usually clearer to split code into another function, rather than assigning an async block to a named variable.

On second though, a join isn't a great solution here, but if it were, that would be better.

Reply:
> > Can you explain why this change would be helpful?
> 
> If a long polling request returns right before a new block is added to the chain, it could delay mining by a few seconds.

Since this code is only supported on testnet, and isn't intended for high-efficiency mining, I think that's ok.

> I think we want to check if that the new template's previous block hash is the same as before,

If you want to do this check to avoid sending useless identical updates, that is implemented here:
https://github.com/ZcashFoundation/zebra/pull/8136/files/1b180e1925762c99d094e3faf536f483d5366590#diff-d73b289b9e87396ee873d64e4bdd5f1d9e57f820adf73df42be03040a3341e89R289

The block hash is the hash of the header, and checking for header equality is actually faster than hashing. (And the block bytes are cryptographically bound to the header, so it's also equivalent to checking for block equality.)

If you want to check if pending solutions should still be sent, that's unlikely to happen, because we submit all solutions as soon as there are one or more solutions from a solver run.

There is a miner hint for sending old blocks in `GetBlockTemplate.submit_old`, and a TODO for it here:
https://github.com/ZcashFoundation/zebra/pull/8136/files/1b180e1925762c99d094e3faf536f483d5366590#diff-d73b289b9e87396ee873d64e4bdd5f1d9e57f820adf73df42be03040a3341e89R438-R440

Since it's rare, and invalid blocks are verified in parallel then correctly rejected, I don't think it's needed.

> or that it's been longer than `GET_BLOCK_TEMPLATE_MEMPOOL_LONG_POLL_INTERVAL` since a new template was sent to the watch channel?

I'm not sure what this means or how it's relevant, sorry!

> > Is the difference important enough to complicate the code?
> 
> I'm hoping it'll be relatively simple, but otherwise no, it's a marginal efficiency gain, and this intended to be used for testing.
> 
> > It is usually clearer to split code into another function, rather than assigning an async block to a named variable.
> 
> On second thought, a join isn't a great solution here, but if it were, that would be better.

A simpler alternative is storing `Instant::now() + GET_BLOCK_TEMPLATE_MEMPOOL_LONG_POLL_INTERVAL` before the RPC call, then using `sleep_until()` here. That could happen in another PR.